client/lib/create-cached-selector/test/index.js

```diff
@@ -0,0 +1,192 @@
+			warn.mockClear();
+			const state = { posts: {} };
+
+			getSitePosts( state, 1 );
+			getSitePosts( state, '' );
+			getSitePosts( state, 'foo' );
+			getSitePosts( state, true );
+			getSitePosts( state, null );
+			getSitePosts( state, undefined );
+			getSitePosts( state, {} );
+			getSitePosts( state, [] );
+			getSitePosts( state, 1, [] );
+
+			expect( warn.mock.calls.length ).toBe( 3 );
+		} );
+
+		test( 'should return the expected value of differing arguments', () => {
+			const state = {
+				posts: {
+					[ post1.id ]: post1,
+					[ post3.id ]: post3,
+				},
+			};
+
+			getSitePosts( state, post1.siteId );
+			const sitePosts = getSitePosts( state, post3.siteId );
+			getSitePosts( state, post1.siteId );
+
+			expect( sitePosts ).toEqual( [ post3 ] );
+			expect( selector.mock.calls.length ).toBe( 2 );
+		} );
+
+		test( 'should bust the cache when watched state changes', () => {
+			const prevState = {
+				posts: {
+					[ post1.id ]: post1,
+				},
+			};
+
+			getSitePosts( prevState, post1.siteId );
+
+			const nextState = {
+				posts: {
+					[ post1.id ]: { ...post1, modified: true },
+					[ post3.id ]: post3,
+				},
+			};
+
+			expect( getSitePosts( nextState, post1.siteId ) ).toEqual( [ { ...post1, modified: true } ] );
+			expect( selector.mock.calls.length ).toBe( 2 );
+		} );
+
+		test( 'should keep the cache on a per-key basis', () => {
+			const getPostByIdWithDataSpy = jest.fn( ( { post } ) => {
+				return {
+					...post,
+					withData: true,
+				};
+			} );
+
+			const getPostByIdWithData = createCachedSelector( {
+				selector: getPostByIdWithDataSpy,
+				getDependents: ( state, postId ) => ( { post: state.posts[ postId ] } ),
+			} );
+
+			const prevState = {
+				posts: {
+					[ post1.id ]: post1,
+				},
+			};
+
+			getPostByIdWithData( prevState, post1.id );
+
+			const nextState = {
+				posts: {
+					[ post1.id ]: post1,
+					[ post2.id ]: post2,
+				},
+			};
+
+			getPostByIdWithData( nextState, post2.id );
+			expect( getPostByIdWithData( nextState, post1.id ) ).toEqual( {
+				...post1,
+				withData: true,
+			} );
+			expect( getPostByIdWithDataSpy.mock.calls.length ).toBe( 2 );
+		} );
+
+		test( 'should throw an error if getDependents is missing', () => {
+			expect( () => createCachedSelector( { selector } ) ).toThrow();
+		} );
+
+		test( 'should throw an error if selector is missing', () => {
+			expect( () => createCachedSelector( { getDependents } ) ).toThrow();
+		} );
+
+		test( 'should call dependant state getter with dependents and arguments', () => {
+			const memoizedSelector = createCachedSelector( { selector, getDependents } );
+			const state = { posts: {} };
+
+			memoizedSelector( state, 1, 2, 3 );
+
+			expect( getDependents ).toHaveBeenCalledWith( { posts: {} }, 1, 2, 3 );
+		} );
+	} );
+
+	describe( '#MixedMap', () => {
+		test( 'should successfully construct a new version', () => {
+			expect( new MixedMap() ).toBeTruthy();
```

Comment:
😏 is this a bit dubious?

`expect( new MixedMap() ).toBeInstanceOf( MixedMap )`

but then again, isn't that a truism? is this test confirming basic JS spec behavior?

```js
function Fu() {}

expect( new Fu() ).toBeTruthy() // passes!
```

Reply:
it does a bit more. it makes sure an exception doesn't get thrown inside the constructor

Reply:
I don't feel strongly. Would you like me to remove it?

Reply:
That seems like a worthwhile and valid test to me. However, we can be explicit about that with something like [`toThrow()`](http://facebook.github.io/jest/docs/en/expect.html#tothrowerror):

```js
expect( () => new Foo() ).not.toThrow()
```

Reply:
Good shout @sirreal! I'll swap it